inbox/util/threading.py

```diff
@@ -16,25 +19,24 @@ def fetch_corresponding_thread(db_session, namespace_id, message):
     # to a message always has a similar subject. This is only
     # right 95% of the time.
     clean_subject = cleanup_subject(message.subject)
+
+    # XXX: It is much faster to sort client-side by message ID, we therefore
+    # use `contains_eager` and `outerjoin` instead of `joinedload` to fetch
+    # the messages by thread in no particular order. We also use a limit to
+    # avoid scanning too many / large threads.
     threads = db_session.query(Thread). \
         filter(Thread.namespace_id == namespace_id,
                Thread._cleaned_subject == clean_subject). \
+        outerjoin(Message, Thread.messages). \
         order_by(desc(Thread.id)). \
         options(load_only('id', 'discriminator'),
-                joinedload(Thread.messages).load_only(
-                    'from_addr', 'to_addr', 'bcc_addr', 'cc_addr'))
+                contains_eager(Thread.messages).load_only(
+                    'from_addr', 'to_addr', 'bcc_addr', 'cc_addr', 'received_date')). \
+        limit(MAX_MESSAGES_SCANNED)
 
-    num_messages_scanned = 0
     for thread in threads:
-        for match in thread.messages:
-            # If we've scanned more than `MAX_MESSAGES_SCANNED`, give up and
-            # assume that a matching thread doesn't exist. We do this because
-            # the number of threads and messages we iterate over can get really
-            # out of hand for some transactional emails.
-            num_messages_scanned += 1
-            if num_messages_scanned > MAX_MESSAGES_SCANNED:
-                return
-
+        messages = sorted(thread.messages, key=attrgetter('received_date'))
```

Comment:
I feel like I'm missing something. Where was this ordering being done before? I don't see it in the `order_by` above.

Reply:
Yes. It used `joinedload` before, which takes the sort from the relation (https://github.com/closeio/sync-engine/blob/master/inbox/models/message.py#L80)